geomstats/geometry/pre_shape.py

```diff
@@ -378,6 +379,339 @@ def integrability_tensor(self, tangent_vec_a, tangent_vec_b, base_point):
+        :math: `\nabla^S_X^X = A_X X`, :math: `\nabla^S_X^Y = A_X Y` \and
+        similarly for Z.
+
+        Parameters
+        ----------
+        tg_vec_x : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`.
+        tg_vec_y : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`.
+        tg_vec_z : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`.
+        base_point : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Point of the total space.
+
+        Returns
+        -------
+        nabla_x_a_y_z : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`, result of :math: `\nabla_X^S
+            (A_Y Z)` with `X = tg_vec_x`, `Y = tg_vec_y` and `Z = tg_vec_z`.
+        a_y_z : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`, result of :math:
+            `A_Y Z` with `Y = tg_vec_y` and `Z = tg_vec_z`.
+
+        References
+        ----------
+        X Pennec. To be published.
+        """
+        # Vectors X and Y have to be horizontal.
+        if not gs.all(self.is_centered(base_point)):
+            raise ValueError('The base_point does not belong to the pre-shape'
+                             ' space')
+        if not gs.all(self.is_horizontal(tg_vec_x, base_point)):
+            raise ValueError('Tangent vector x is not horizontal')
+        if not gs.all(self.is_horizontal(tg_vec_y, base_point)):
+            raise ValueError('Tangent vector y is not horizontal')
+        if not gs.all(self.is_horizontal(tg_vec_z, base_point)):
+            raise ValueError('Tangent vector z is not horizontal')
+        # hor_x = self.horizontal_projection(tangent_vec_x, base_point)
+        # hor_y = self.horizontal_projection(tangent_vec_y, base_point)
+
+        p_top = Matrices.transpose(base_point)
+        p_top_p = gs.matmul(p_top, base_point)
+
+        def sylv_p(mat_b):
+            return gs.linalg.solve_sylvester(
+                p_top_p, p_top_p, mat_b - Matrices.transpose(mat_b))
+
+        # \Omega_{YZ}  = \Sylv_P(Z^\top Y)
+        z_top = Matrices.transpose(tg_vec_z)
+        y_top = Matrices.transpose(tg_vec_y)
+        omega_yz = sylv_p(gs.matmul(z_top, tg_vec_y))
+        a_y_z = gs.matmul(base_point, omega_yz)
+        omega_xy = sylv_p(gs.matmul(y_top, tg_vec_x))
+        omega_xz = sylv_p(gs.matmul(z_top, tg_vec_x))
+
+        omega_yz_x = gs.matmul(tg_vec_x, omega_yz)
+        omega_xz_y = gs.matmul(tg_vec_y, omega_xz)
+        omega_xy_z = gs.matmul(tg_vec_z, omega_xy)
+
+        tg_vec_f = 2. * omega_yz_x + omega_xz_y - omega_xy_z
+        omega_fp = sylv_p(gs.matmul(p_top, tg_vec_f))
+        omega_fp_p = gs.matmul(base_point, omega_fp)
+
+        nabla_x_a_y_z = omega_yz_x - omega_fp_p
+
+        return nabla_x_a_y_z, a_y_z
+
+    def iterated_integrability_tensor_derivative_parallel(
```

Comment:
These are not the same functions. the (non-iterated) special case computes `\nabla_X (A_Y Z)` with horizontal vectors X, Y and Z, while the iterated version computes `\nabla_X (A_Y A_X Y)` (the integrability tensor is used twice here). Although this could be seen as `\nabla_X (A_Y V)`, the vertical vector V need to have the very special form `V = A_X Y` where X and Y are horizontal vectors. 

Merging both functions under the same hat would bring more confusion than clarification, especially since the signature differs.